### PR TITLE
hwloop: Fix chain dependency bug for hwloop instr

### PR DESCRIPTION
**Pull request by kanishkan**
If there is a node between hwloop and br, the pattern
matcher fails to select hwloop instructions. This fix
manually moves the hwloop instrution close to br by
updating chain dependencies manually.

### CODE REVIEW

**Comment by kanishkan** [comment]
Hello @pjaaskel, this is ready for review.

File: tce/opset/base/base.cc [hunk 1]
```diff
@@ -2334,12 +2334,15 @@ END_DEFINE_STATE
 OPERATION_WITH_STATE(HWLOOP, HLOOP_SETUP)
 TRIGGER
     // sanity check
-    if (STATE.hwloopEnabled) {
-        RUNTIME_ERROR("Already hwloop is in progress");
```
Comment (pjaaskel) [inline tce/opset/base/base.cc:2338]: -> Hardware loop execution already in progress.

File: tce/opset/base/base.cc [hunk 2]
```diff
@@ -2334,12 +2334,15 @@ END_DEFINE_STATE
...
-    }
     if (UINT(1) == 0) {
         RUNTIME_ERROR("hwloop for iterations=0 is invalid");
```
Comment (pjaaskel) [inline tce/opset/base/base.cc:2338]: -> Hardware loop invocation with 0 iterations is invalid.

File: tce/opset/base/base.cc [hunk 3]
```diff
@@ -2334,12 +2334,15 @@ END_DEFINE_STATE
...
     }
+    if (UINT(2) == 0) {
+        RUNTIME_ERROR("hwloop for instr-length=0 is invalid");
```
Comment (pjaaskel) [inline tce/opset/base/base.cc:2341]: ...with instruction length 0 is invalid.

File: tce/opset/base/base.cc [hunk 4]
```diff
@@ -2334,12 +2334,15 @@ END_DEFINE_STATE
...
+    }
+    if (STATE.hwloopEnabled) {
```
Comment (pjaaskel) [inline tce/opset/base/base.cc:2343]: Duplication?

File: tce/src/applibs/LLVMBackend/TDGen.cc [hunk 5]
```diff
@@ -3474,7 +3474,7 @@ TDGen::writeInstrInfo(std::ostream& os) {
     writeControlFlowInstrDefs(os);
 
     // Hardware loop instructions
-    if (mach_.controlUnit()->hasOperation("hwloop")) writeHWLoopDef(os);
+    writeHWLoopDef(os);
```
Comment (pjaaskel) [inline tce/src/applibs/LLVMBackend/TDGen.cc:3477]: Is this on purpose? It now always writes the HWLoopDef even if the ADF doesn't have it?

File: tce/src/applibs/Scheduler/ProgramRepresentations/CFG/BasicBlockNode.hh [hunk 6]
```diff
@@ -102,6 +102,16 @@ public:
     void setLoopScheduled() { loopScheduled_ = true; }
     void setBBOwnership(bool ownership = true) { bbOwned_ = ownership; }
 
+    // hwloop flags
+    bool
+    isHWLoop() const {
```
Comment (pjaaskel) [inline tce/src/applibs/Scheduler/ProgramRepresentations/CFG/BasicBlockNode.hh:107]: Please add a more thorough comment describing what are the conditions when this is set to true.

File: tce/src/applibs/Scheduler/ProgramRepresentations/CFG/BasicBlockNode.hh [hunk 7]
```diff
@@ -102,6 +102,16 @@ public:
...
+        return hwloop_;
+    }
+    void
+    setHWLoop() {
```
Comment (pjaaskel) [inline tce/src/applibs/Scheduler/ProgramRepresentations/CFG/BasicBlockNode.hh:111]: add a bool parameter so it's possible to unset?

File: tce/src/base/program/BasicBlock.hh [hunk 8]
```diff
@@ -108,6 +108,16 @@ public:
     unsigned tripCount() const { return tripCount_; }
     void setTripCount(unsigned count) { tripCount_ = count; }
 
+    /// BB is known to be a hwloop body
```
Comment (pjaaskel) [inline tce/src/base/program/BasicBlock.hh:111]: Why duplicate this in BasicBlockNode? Why not call the BB method? Is it needed outside scheduling (thus does BasicBlock need it really or BasicBlockNode would be enough)?

File: tce/src/base/program/BasicBlock.hh [hunk 9]
```diff
@@ -117,6 +127,7 @@ private:
     BasicBlockStatistics statistics_;    
 
     bool innerLoop_;
+    bool hwLoop_;
```
Comment (pjaaskel) [inline tce/src/base/program/BasicBlock.hh:130]: isHardwareLoop_;

**Review by pjaaskel — CHANGES_REQUESTED** [review]
Thanks! Please check a couple of minor issues.

File: tce/src/base/program/BasicBlock.hh (hunk 9 above)
Comment (kanishkan) [reply tce/src/base/program/BasicBlock.hh:130]: Updated

File: tce/opset/base/base.cc (hunk 1 above)
Comment (kanishkan) [reply tce/opset/base/base.cc:2338]: Done

File: tce/opset/base/base.cc (hunk 2 above)
Comment (kanishkan) [reply tce/opset/base/base.cc:2338]: done

File: tce/opset/base/base.cc (hunk 4 above)
Comment (kanishkan) [reply tce/opset/base/base.cc:2343]: We have two parameters. Loop length and iteration count. We expect both to be >0.

File: tce/src/applibs/LLVMBackend/TDGen.cc (hunk 5 above)
Comment (kanishkan) [reply tce/src/applibs/LLVMBackend/TDGen.cc:3477]: Inside `writeHWLoopDef` we have a condition to check "hwloop" instruction in the adf.
However, we have to add `LJUMP` instruction to TDGen irrespective of `hwloop` instruction. 

We are doing custom instr-selection for LJUMP. So, we need to hard-code its opcode (TCE::LJUMP) in the TCEISelector.
That is why we moved this "hwloop" check inside writeHWLoopDef().

File: tce/src/applibs/Scheduler/ProgramRepresentations/CFG/BasicBlockNode.hh (hunk 7 above)
Comment (kanishkan) [reply tce/src/applibs/Scheduler/ProgramRepresentations/CFG/BasicBlockNode.hh]: Done

File: tce/src/base/program/BasicBlock.hh (hunk 8 above)
Comment (kanishkan) [reply tce/src/base/program/BasicBlock.hh]: You are right! We can get away with just having this on BasicBlockNode. I have updated the patch for it.

**Comment by kanishkan** [comment]
@pjaaskel : Thanks for your review. Resolved the comments. Let me know if any other changes are need.

**Review by pjaaskel — APPROVED** [review]

**Pull request merged**